### PR TITLE
Tags/db schema

### PR DESCRIPTION
**Pull request by filafb**
# New Feature: Tags 🎉

## Description

Schema representation: https://dbdiagram.io/d/5e72b4424495b02c3b887056

This PR includes the migration file - up and down methods - for two new tables and their models representation. I included a few unit tests that, when we have GraphQL schema and resolvers, we can use/adapt them as integrated tests.

Why is longer than 300 lines? Test setup and 4 brand new files.

# Checklist:

- [ ] I have manually tested my changes on desktop and mobile
- [x] The test suite passes locally with my changes
- [ ] If my change is a UI change, I have attached a screenshot to the description section of this pull request
- [x] [My change is 300 lines of code or less](https://github.com/MoveOnOrg/Spoke/blob/main/CONTRIBUTING.md#your-first-code-contribution), or has a documented reason in the description why it’s longer
- [ ] I have made any necessary changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] My PR is labeled [WIP] if it is in progress

### CODE REVIEW

File: migrations/20200322171627_tags.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        t.text("description").notNullable();
+        t.boolean("is_deleted").defaultTo(false);
+        t.timestamp("created_at")
+          .notNullable()
+          .defaultTo(knex.fn.now());
+        t.timestamp("updated_at")
+          .notNullable()
+          .defaultTo(knex.fn.now());
+        t.integer("organization_id").notNullable();
+        t.foreign("organization_id").references("organization.id");
```
Comment (schuyler1d) [inline migrations/20200322171627_tags.js]: can you wrap the `.foreign` call in a `if (!isSqlite) {` (see init_db.js file for examples and what that variable is) -- annoying, but necessary

File: migrations/20200322171627_tags.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    },
+    {
+      tableName: "tag_campaign_contact",
+      create: t => {
+        t.increments("id").primary();
+        t.text("value");
+        t.integer("tag_id").notNullable();
+        t.foreign("tag_id").references("tag.id");
+        t.integer("campaign_contact_id").notNullable();
+        t.foreign("campaign_contact_id").references("campaign_contact.id");
```
Comment (schuyler1d) [inline migrations/20200322171627_tags.js]: same for all the .foreign here. (!isSqlite)

File: migrations/20200322171627_tags.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      const newTable = await knex.schema.createTable(tableName, create);
+      tablePromises.push(newTable);
+    }
+  }
+
+  return Promise.all(tablePromises);
+};
+
+exports.down = knex => {
+  return knex.schema.dropTableIfExists("tag_content").dropTableIfExists("tag");
```
Comment (schuyler1d) [inline migrations/20200322171627_tags.js]: this needs to be `tag_campaign_contact` as above, right?

File: src/server/models/index.js [hunk 4]
```diff
@@ -59,7 +61,9 @@ const tableList = [
   "question_response",
   "user_cell",
   "user_organization",
-  "zip_code" // good candidate (or by contact)?
+  "zip_code", // good candidate (or by contact)?,
+  "tag",
```
Comment (schuyler1d) [inline src/server/models/index.js]: please move to alphabetical (just above user_cell, I guess)

File: migrations/20200322171627_tags.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+      }
+    },
+    {
+      tableName: "tag_campaign_contact",
+      create: t => {
+        t.increments("id").primary();
+        t.text("value");
+        t.integer("tag_id").notNullable();
+        t.foreign("tag_id").references("tag.id");
+        t.integer("campaign_contact_id").notNullable();
```
Comment (schuyler1d) [inline migrations/20200322171627_tags.js:33]: Please create an index on campaign_contact_id.

File: migrations/20200322171627_tags.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        t.text("name").notNullable();
+        t.text("description").notNullable();
+        t.boolean("is_deleted").defaultTo(false);
+        t.timestamp("created_at")
+          .notNullable()
+          .defaultTo(knex.fn.now());
+        t.timestamp("updated_at")
+          .notNullable()
+          .defaultTo(knex.fn.now());
+        t.integer("organization_id").notNullable();
```
Comment (schuyler1d) [inline migrations/20200322171627_tags.js:19]: Please create an index on organization_id (bonus for a single two-field index of [organization_id, is_deleted])

File: migrations/20200322171627_tags.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
+exports.up = async knex => {
+  const buildTableSchema = [
+    {
+      tableName: "tag",
+      create: t => {
+        t.increments("id").primary();
+        t.text("group");
+        t.text("name").notNullable();
+        t.text("description").notNullable();
```
Comment (schuyler1d) [inline migrations/20200322171627_tags.js:11]: is description 'friendly name' or is it a long detailed field?  If it's the display name, maybe call it either `title` or `display_name`?  description is fine, too, but then let's add some comments that describe what it's for (maybe in the models/tag.js file)

File: __test__/backend.test.js [hunk 8]
```diff
@@ -805,3 +824,185 @@ describe("editUser mutation", () => {
     });
   });
 });
+
+describe("A tag table", () => {
```
Comment (schuyler1d) [inline __test__/backend.test.js]: please move these to a new file -- we don't add anything to backend.test.js anymore.  how about `__test__/server/models/tag.js`? or server/tag.js, especially if you plan on populating more later with some more application-y code.

**Review by schuyler1d — CHANGES_REQUESTED** [review]
Off to the races!!!!!!  :fireworks: :rocket:

File: __test__/backend.test.js (hunk 8 above)
Comment (filafb) [reply __test__/backend.test.js]: Done!

File: migrations/20200322171627_tags.js (hunk 5 above)
Comment (filafb) [reply migrations/20200322171627_tags.js:33]: Just added it but not sure how to test index. Any suggestion?

File: migrations/20200322171627_tags.js (hunk 6 above)
Comment (filafb) [reply migrations/20200322171627_tags.js:19]: Done! I haven't worked with indexes before. Worth a double check.

File: migrations/20200322171627_tags.js (hunk 7 above)
Comment (filafb) [reply migrations/20200322171627_tags.js:11]: I am not sure. The documentation named it as description. Open to modify it if needed.

Comment (schuyler1d) [reply migrations/20200322171627_tags.js:11]: ok. let's leave it an wait until we have an interface.

File: migrations/20200322171627_tags.js (hunk 5 above)
Comment (schuyler1d) [reply migrations/20200322171627_tags.js:33]: you shouldn't have to test the index.  generally we want to test application features/code -- testing the schema itself isn't as important.

**Review by schuyler1d — APPROVED** [review]
:+1:

**Pull request merged**